src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,441 @@
+            {
+                if (this.cachedHeaders.Count != 0)
+                {
+                    result = await this.PresentHeadersLockedAsync(this.cachedHeaders, false).ConfigureAwait(false);
+
+                    if (result == null)
+                    {
+                        this.logger.LogTrace("(-)[NO_HEADERS_CONNECTED]:null");
+                        return null;
+                    }
+
+                    this.ExpectedPeerTip = result.Consumed;
+
+                    int consumedCount = this.cachedHeaders.IndexOf(result.Consumed.Header) + 1;
+                    this.cachedHeaders.RemoveRange(0, consumedCount);
+                    int cacheSize = this.cachedHeaders.Count;
+
+                    this.logger.LogTrace("{0} entries were consumed from the cache, {1} items were left.", consumedCount, cacheSize);
+                    syncRequired = cacheSize == 0;
+                }
+            }
+
+            if (syncRequired)
+                await this.ResyncAsync().ConfigureAwait(false);
+
+            this.logger.LogTrace("(-):'{0}'", result);
+            return result;
+        }
+
+        /// <summary>
+        /// Processes and incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        private async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(message), message.Message.Command);
+
+            switch (message.Message.Payload)
+            {
+                case GetHeadersPayload getHeaders:
+                    await this.ProcessGetHeadersAsync(peer, getHeaders).ConfigureAwait(false);
+                    break;
+
+                case HeadersPayload headers:
+                    await this.ProcessHeadersAsync(peer, headers).ConfigureAwait(false);
+                    break;
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// Processes "getheaders" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="getHeadersPayload">Payload of "getheaders" message to process.</param>
+        /// <remarks>
+        /// "getheaders" message is sent by the peer in response to "headers" message until an empty array is received.
+        /// <para>
+        /// This payload notifies peers of our current best validated height, which is held by consensus tip.
+        /// </para>
+        /// <para>
+        /// If the peer is behind/equal to our best height an empty array is sent back.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessGetHeadersAsync(INetworkPeer peer, GetHeadersPayload getHeadersPayload)
```

Comment:
@noescape00 thanks for making the changes thus far, it looks a lot better. However, this method along with `ProcessHeadersAsync` violates SRP (Single Responsilibity Principle). To me, this method does 3 things;

1. Check to see that we are in IBD
2. Construct a list of headers to send
3. Send headers to the peer

This in itself warrants 3 separate  methods to handle individual logic as SRP states that a class or method should only ever be responsible for one thing and one thing only.

Its either that or the consumer of this method (`OnMessageReceived)` should do the IBD check upfront before calling the method. But the piece of logic that creates the set of headers should definitly be separate piece of logic that can be tested on its own.

Reply:
The IBD check does not count because it is a guard.

Reply:
Fixed

Reply:
@Aprogiena good point on the Guard! @noescape00 good job, this method looks way better now 👍  Getting there!